packages/solana/src/utils/wallet-standard/adapter.ts

```diff
@@ -0,0 +1,548 @@
+        try {
+          await this.#wallet.features[StandardConnect].connect(input)
+        } catch (error: unknown) {
+          throw new WalletConnectionError((error as Error)?.message, error)
+        }
+      }
+
+      const account = this.#wallet.accounts[0]
+      if (!account) {
+        throw new WalletAccountError()
+      }
+
+      this.#connected(account)
+    } catch (error: unknown) {
+      this.emit('error', error as WalletError)
+      throw error
+    } finally {
+      this.#connecting = false
+    }
+  }
+
+  async disconnect(): Promise<void> {
+    if (StandardDisconnect in this.#wallet.features) {
+      try {
+        this.#disconnecting = true
+        await this.#wallet.features[StandardDisconnect].disconnect()
+      } catch (error: unknown) {
+        this.emit('error', new WalletDisconnectionError((error as Error)?.message, error))
+      } finally {
+        this.#disconnecting = false
+      }
+    }
+
+    this.#disconnected()
+  }
+
+  #connected(account: WalletAccount) {
+    let publicKey: PublicKey | undefined = undefined
+    try {
+      // Use account.address instead of account.publicKey since address could be a PDA
+      publicKey = new PublicKey(account.address)
+    } catch (error: unknown) {
+      throw new WalletPublicKeyError((error as Error)?.message, error)
+    }
+
+    this.#account = account
+    this.#publicKey = publicKey
+    this.#reset()
+    this.emit('connect', publicKey)
+  }
+
+  #disconnected(): void {
+    this.#account = null
+    this.#publicKey = null
+    this.#reset()
+    this.emit('disconnect')
+  }
+
+  #reset() {
+    const supportedTransactionVersions =
+      SolanaSignAndSendTransaction in this.#wallet.features
+        ? this.#wallet.features[SolanaSignAndSendTransaction].supportedTransactionVersions
+        : this.#wallet.features[SolanaSignTransaction].supportedTransactionVersions
+    this.#supportedTransactionVersions = arraysEqual(supportedTransactionVersions, ['legacy'])
+      ? null
+      : new Set(supportedTransactionVersions)
+
+    if (
+      SolanaSignTransaction in this.#wallet.features &&
+      this.#account?.features.includes(SolanaSignTransaction)
+    ) {
+      this.signTransaction = this.#signTransaction
+      this.signAllTransactions = this.#signAllTransactions
+    } else {
+      delete this.signTransaction
+      delete this.signAllTransactions
+    }
+
+    if (
+      SolanaSignMessage in this.#wallet.features &&
+      this.#account?.features.includes(SolanaSignMessage)
+    ) {
+      this.signMessage = this.#signMessage
+    } else {
+      delete this.signMessage
+    }
+
+    if (SolanaSignIn in this.#wallet.features) {
+      this.signIn = this.#signIn
+    } else {
+      delete this.signIn
+    }
+  }
+
+  #changed: StandardEventsListeners['change'] = properties => {
+    // If accounts have changed on the wallet, reflect this on the adapter.
+    if ('accounts' in properties) {
+      const account = this.#wallet.accounts[0]
```

Comment:
Why we are checking the first account every time? Let's say there are 3 accounts and 1 is connected. Is the connected one always in the first index?

Reply:
Yes, when account is switched, wallet.accounts has only one element - with info about active account